src/storage/in_mem_csv_copier/in_mem_arrow_node_copier.cpp

```diff
@@ -0,0 +1,626 @@
+            }
+        }
+
+        taskScheduler.waitAllTasksToCompleteOrError();
+        return arrow::Status::OK();
+    }
+
+    template<typename T>
+    arrow::Status InMemArrowNodeCopier::populateColumnsFromParquet(unique_ptr<HashIndexBuilder<T>> &pkIndex) {
+        node_offset_t offsetStart = 0;
+
+        std::unique_ptr<parquet::arrow::FileReader> reader;
+        auto status = initParquetReader(reader, csvDescription.filePath);
+        if (!status.ok()) {
+            return status;
+        }
+
+        std::shared_ptr<arrow::Table> currTable;
+        int blockIdx = 0;
+        while (blockIdx < numBlocks) {
+            for (int batchOfTasksId = 0; batchOfTasksId < CopyCSVConfig::NUM_TASKS_PER_BATCH; ++ batchOfTasksId) {
+                if (blockIdx == numBlocks) {
+                    break;
+                }
+                ARROW_RETURN_NOT_OK(reader->RowGroup(blockIdx)->ReadTable(&currTable));
+                taskScheduler.scheduleTask(CopyCSVTaskFactory::createCopyCSVTask(
+                        batchPopulateColumnsTask<T, arrow::ChunkedArray>,
+                        nodeTableSchema->primaryKeyPropertyIdx,
+                        blockIdx, offsetStart, pkIndex.get(), this, currTable->columns(),
+                        csvDescription.csvReaderConfig.tokenSeparator));
+                offsetStart += currTable->num_rows();
+                ++ blockIdx;
+            }
+            while (taskScheduler.getTaskNum() > CopyCSVConfig::MINIMUM_TASKS_TO_SCHEDULE_MORE) {
+                taskScheduler.errorIfThereIsAnException();
+                this_thread::sleep_for(chrono::microseconds(THREAD_SLEEP_TIME_WHEN_WAITING_IN_MICROS));
+            }
+        }
+
+        taskScheduler.waitAllTasksToCompleteOrError();
+
+        return arrow::Status::OK();
+    }
+
+    template<typename T>
+    void InMemArrowNodeCopier::addIDsToIndex(InMemColumn *column, HashIndexBuilder<T> *hashIndex,
+                                             node_offset_t startOffset, uint64_t numValues) {
+        for (auto i = 0u; i < numValues; i++) {
+            auto offset = i + startOffset;
+            if constexpr (is_same<T, int64_t>::value) {
+                auto key = (int64_t *) column->getElement(offset);
+                if (!hashIndex->append(*key, offset)) {
+                    throw CopyCSVException(Exception::getExistedPKExceptionMsg(to_string(*key)));
+                }
+            } else {
+                auto element = (ku_string_t *) column->getElement(offset);
+                auto key = column->getInMemOverflowFile()->readString(element);
+                if (!hashIndex->append(key.c_str(), offset)) {
+                    throw CopyCSVException(Exception::getExistedPKExceptionMsg(key));
+                }
+            }
+        }
+    }
+
+    template<typename T>
+    void InMemArrowNodeCopier::populatePKIndex(InMemColumn *column, HashIndexBuilder<T> *pkIndex,
```

Comment:
This function is no longer justified because it simply calls another function. Either move the contents of addIDsToIndex to here and remove addIDsToIndex. Or remove this function and only keep addIDsToIndex.